### PR TITLE
Bump PHPStan level to 2

### PR DESCRIPTION
**Pull request by phansys**
## Subject
Bump PHPStan level to 2.

I am targeting this branch, because these changes respect BC.

## Changelog

```markdown
### Changed
- Several missing or wrong type hints.
```

### CODE REVIEW

**Comment by phansys** [comment]
>If you up phpstan levels or add phpdoc please use master for reference and help with the merge after the Pr gets accepted. It could cause multiple conflicts

I need these changes in the 3.x branch.
Have we updated our policies about this?

**Comment by jordisala1991** [comment]
> >If you up phpstan levels or add phpdoc please use master for reference and help with the merge after the Pr gets accepted. It could cause multiple conflicts
> 
> 
> 
> I need these changes in the 3.x branch.
> 
> Have we updated our policies about this?

You can do them on 3.x, nothing has changed.
What Im telling is that those changes are already on master probably, take a look and backport what you need, to avoid unnecesary conflicts.

**Comment by phansys** [comment]
Thank you @jordisala1991, When this PR is merged, I'll try to perform the merge against `master`.

File: src/Admin/BaseMediaAdmin.php [hunk 1]
```diff
@@ -53,6 +60,11 @@ public function __construct($code, $class, $baseControllerName, Pool $pool, ?Cat
         $this->categoryManager = $categoryManager;
     }
 
+    /**
+     * @param MediaInterface $object
+     *
+     * @phpstan-param T $object
+     */
```
Comment (VincentLanglet) [inline src/Admin/BaseMediaAdmin.php:67]: Why is this needed, it should be in the parent interface, isn't it ?

File: src/Admin/BaseMediaAdmin.php [hunk 2]
```diff
@@ -103,6 +115,11 @@ public function getPersistentParameters()
         ]);
     }
 
+    /**
+     * @return MediaInterface
+     *
+     * @phpstan-return T
+     */
```
Comment (VincentLanglet) [inline src/Admin/BaseMediaAdmin.php:122]: same

File: src/Admin/BaseMediaAdmin.php [hunk 3]
```diff
@@ -138,6 +155,11 @@ public function getPool()
         return $this->pool;
     }
 
+    /**
+     * @param MediaInterface $object
+     *
+     * @phpstan-param T $object
+     */
```
Comment (VincentLanglet) [inline src/Admin/BaseMediaAdmin.php:162]: same

File: src/Admin/BaseMediaAdmin.php [hunk 4]
```diff
@@ -161,13 +183,9 @@ protected function configureListFields(ListMapper $list)
 
     protected function configureFormFields(FormMapper $form)
     {
-        $media = $this->getSubject();
-
-        if (!$media) {
-            $media = $this->getNewInstance();
-        }
+        $media = $this->hasSubject() ? $this->getSubject() : $this->getNewInstance();
```
Comment (VincentLanglet) [inline src/Admin/BaseMediaAdmin.php:186]: I'm not sure it makes a lot of sens since SonataAdmin have a check to throw an error if there is no subject when building the form/show.

File: src/Command/AddMassMediaCommand.php [hunk 5]
```diff
@@ -120,7 +122,10 @@ protected function insertMedia(array $data, OutputInterface $output)
     protected function optimize()
     {
         if (null !== $this->managerRegistry) {
-            $this->managerRegistry->getManager()->getUnitOfWork()->clear();
+            $manager = $this->managerRegistry->getManager();
+
+            \assert($manager instanceof DocumentManagerInterface || $manager instanceof EntityManagerInterface);
+            $manager->getUnitOfWork()->clear();
```
Comment (VincentLanglet) [inline src/Command/AddMassMediaCommand.php:128]: I think we introduce a ClearableManagerInterface, couldn't we do an instanceof check instead ?

https://github.com/sonata-project/sonata-doctrine-extensions/pull/323/files

File: src/Command/BaseCommand.php [hunk 6]
```diff
@@ -29,7 +30,7 @@
 abstract class BaseCommand extends Command
 {
     /**
-     * @var ManagerInterface
+     * @var ManagerInterface&ClearableManagerInterface
```
Comment (VincentLanglet) [inline src/Command/BaseCommand.php:33]: I'm not sure the `ClearableManagerInterface` is required.

When using `clear, we should just make an `instanceof` check before calling clear.
This is what we did on master.

File: src/Command/UpdateCdnStatusCommand.php [hunk 7]
```diff
@@ -84,6 +85,9 @@ public function execute(InputInterface $input, OutputInterface $output)
         $this->output = $output;
 
         $provider = $this->getProvider();
+
+        \assert($provider instanceof BaseProvider);
```
Comment (VincentLanglet) [inline src/Command/UpdateCdnStatusCommand.php:89]: What the purpose of this assert ?

I prefer to understand because it often show an issue
- with our return type
- with the use of the return values
- with our interfaces

Comment (phansys) [reply src/Command/UpdateCdnStatusCommand.php:89]: I need to address #1824 in order to fix this comment.

File: src/Admin/BaseMediaAdmin.php (hunk 4 above)
Comment (phansys) [reply src/Admin/BaseMediaAdmin.php:186]: I agree, but if we throw an exception in this case, it could be considered a BC-break, since it isn't the current behavior.

Comment (VincentLanglet) [reply src/Admin/BaseMediaAdmin.php:180]: Since there is the exception for 4.0 https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Admin/AbstractAdmin.php#L896 and a deprecation for 3.x, I would not add the `hasSubject` check here.

And we could add a next major to explain that in 4.0, `$media = $this->getSubject()` will be enough, without the null check.

File: src/Command/UpdateCdnStatusCommand.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Command/UpdateCdnStatusCommand.php:89]: Can we have a comment with the ref of the issue then ? (just in case)

File: src/Admin/BaseMediaAdmin.php (hunk 4 above)
Comment (phansys) [reply src/Admin/BaseMediaAdmin.php:180]: I think I don't get your point, as `$media` is being used several times in this method trusting it is always an instance of `MediaInterface`.

File: src/Command/UpdateCdnStatusCommand.php (hunk 7 above)
Comment (phansys) [reply src/Command/UpdateCdnStatusCommand.php]: I finally added the `@method` annotation at `CDNIterface`.

File: src/Admin/BaseMediaAdmin.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Admin/BaseMediaAdmin.php:180]: In next major
```
$media = $this->hasSubject() ? $this->getSubject() : $this->getNewInstance();
```
can be replaced to
```
$media = $this->getSubject();
```

Comment (phansys) [reply src/Admin/BaseMediaAdmin.php:180]: I agree that is the proper behavior for the next major, but I don't understand what are you suggesting here for the 3.x branch, where `$this->getSubject()` returns `null`. Could you please provide a suggestion?

**Review by VincentLanglet — COMMENTED** [review]
Honestly I feel like we're passing a lot of time on this.
- The PR is huge, might be easier to create smaller pr fixing 5-10 files at the times.
- The conflict on the merge will be hell
- PHPStan errors should be solved the same way we did on master, we already had a lot of reflexion and discussion about the best way to do it.

BTW, by adding `@method`, the label is not pedantic anymore.

You said you "need" this for your project. I kinda curious why, and if we couldn't just fix the errors needed for your project and keep working on releasing master as fast as possible.
- Phpstan is already lvl max
- Psalm will be lvl 3: https://github.com/sonata-project/SonataMediaBundle/pull/2038
- Support for php8 was added
- Support for Symfony 5 was added
- I dunno if everything is final and type hinted, i need to check

Do you have a list of what is missing for v4 @jordisala1991 ?

File: src/Admin/BaseMediaAdmin.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Admin/BaseMediaAdmin.php:180]: I just thought about adding a `Next_major` comment in order to not forget to do the change during the merge into master.

```
// NEXT_MAJOR: use the next line instead.
$media = $this->hasSubject() ? $this->getSubject() : $this->getNewInstance();
// $media = $this->getSubject();
```

**Comment by phansys** [comment]
>The PR is huge, might be easier to create smaller pr fixing 5-10 files at the times.

I'll try to provide parts of these changes in multiple PRs. In fact, all the PRs I created the last 2 weeks in this project were extracted from the findings in this PR.

>The conflict on the merge will be hell

I know, I'll perform the merge when this PR is merged.

>PHPStan errors should be solved the same way we did on master, we already had a lot of reflexion and discussion about the best way to do it.
BTW, by adding @method, the label is not pedantic anymore.

I'll try to check for the changes done in the `master` branch again.

>You said you "need" this for your project. I kinda curious why, and if we couldn't just fix the errors needed for your project and keep working on releasing master as fast as possible.

The majority of errors reported by PHPStan in my projects are generated by the lack of some methods in the interfaces (which are only present in their implementations), and by covariance issues regarding the implementation of these interfaces (in some cases, from my project's classes which also extend or implement interfaces declared here).
I must check if I'm able to identify these specific cases and provide a narrower PR for them.

**Comment by core23** [comment]
Can you rebase this @phansys ?

File: phpstan.neon.dist [hunk 8]
```diff
@@ -2,14 +2,14 @@ includes:
...
+    level: 2
     bootstrapFiles:
         - vendor/bin/.phpunit/phpunit/vendor/autoload.php
-    symfony:
-        console_application_loader: ./phpstan-console-application.php
     scanFiles:
         # NEXT_MAJOR: Remove the scanFiles part
         - .phpstan/stubs/MetadataInterface.file
+    symfony:
+        console_application_loader: ./phpstan-console-application.php
```
Comment (ChristianGripp) [inline phpstan.neon.dist]: Maybe move this file to `.phpstan/`

Comment (VincentLanglet) [reply phpstan.neon.dist]: We tried to not use a .phpstan folder for others bundles

Comment (ChristianGripp) [reply phpstan.neon.dist]: Then we should exclude the file via `.gitattributes`

Comment (VincentLanglet) [reply phpstan.neon.dist]: https://github.com/sonata-project/dev-kit/blob/master/templates/project/.gitattributes.twig#L16

File: src/PHPCR/BaseGallery.php [hunk 9]
```diff
@@ -44,10 +45,12 @@ public function getUuid()
 
     public function addGalleryHasMedia(GalleryHasMediaInterface $galleryHasMedia)
     {
+        \assert($galleryHasMedia instanceof BaseGalleryHasMedia);
```
[the remark this answers is not in the document]
Comment (phansys) [reply src/PHPCR/BaseGallery.php:48]: I don't understand your comment here. Can you please provide a suggestion?

**Review by jordisala1991 — DISMISSED** [review]
It is nearly impossible to review this PR, all of this changes and more are done already on 4.x and I can't go one by one. If you are sure about doing this, you have my approval, but you merge this one on 4.x. Keep in mind that this should produce 0 changes on 4.x, if it produces any it means that something here is wrong (most likely, because on 4.x we have psalm level 2 and phpstan level 8 checking the code)

**Review by VincentLanglet — DISMISSED** [review]
I agree with @jordisala1991

**Comment by phansys** [comment]
>Keep in mind that this should produce 0 changes on 4.x, if it produces any it means that something here is wrong (most likely, because on 4.x we have psalm level 2 and phpstan level 8 checking the code)

I'll try to find some time to perform this check in order to confirm the consistency.

**Comment by VincentLanglet** [comment]
> I'll try to find some time to perform this check in order to confirm the consistency.

I would recommend to merge this and then merge 3.x into 4.x, the conflict should show the differences.

**Comment by phansys** [comment]
>I would recommend to merge this and then merge 3.x into 4.x, the conflict should show the differences.

Sure, that check is what I have in mind.

File: phpstan-baseline.neon [hunk 10]
```diff
@@ -9,3 +9,28 @@ parameters:
...
+
+        -
+            message: "#^Call to an undefined method Sonata\\\\ClassificationBundle\\\\Model\\\\CategoryManagerInterface\\:\\:getRootCategory\\(\\)\\.$#"
+            count: 1
+            path: src/Model/CategoryManager.php
+
+        -
+            message: "#^Call to an undefined method Twig\\\\Environment\\:\\:expects\\(\\)\\.$#"
+            count: 1
+            path: tests/Block/GalleryBlockServiceTest.php
```
Comment (ChristianGripp) [inline phpstan-baseline.neon]: We should fix this error here: 
https://github.com/sonata-project/SonataBlockBundle/blob/3.x/src/Test/BlockServiceTestCase.php#L50-L53

**Review by VincentLanglet — DISMISSED** [review]

**Review by VincentLanglet — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
@phansys Are you available to merge this, and do the 3.x into 4.x merge (to check).
Then we'll release the next 3.x version.

**Comment by phansys** [comment]
>@phansys Are you available to merge this, and do the 3.x into 4.x merge (to check).
Then we'll release the next 3.x version.

Sorry for the delay @VincentLanglet. I'll try to find time this week.

File: src/Model/GalleryInterface.php [hunk 11]
```diff
@@ -93,23 +83,29 @@ public function getCreatedAt();
     public function setDefaultFormat($defaultFormat);
```
Comment (ChristianGripp) [inline src/Model/GalleryInterface.php:93]: You can add `@return void` to all setters

**Review by core23 — DISMISSED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by phansys** [comment]
Friendly ping @jordisala1991, @core23.
3.37.0 (#2218) will be released soon and I like to include these changes there.

**Review by jordisala1991 — APPROVED** [review]

**Comment by core23** [comment]
There are a lot of merge conflicts. Can you try to merge this into the 4.x branch?
